app/app.go

```diff
@@ -168,8 +174,52 @@ func (app *App) InitChain(req abci.RequestInitChain) (res abci.ResponseInitChain
 	return
 }
 
-// Implements ABCI
+// Query implements ABCI and allows use to externally inspect state
 func (app *App) Query(req abci.RequestQuery) (res abci.ResponseQuery) {
+	// TODO: actually implement this
+	if req.Prove {
+		return abci.ResponseQuery{Code: 500, Log: "Query with proof not implemented"}
```

Comment:
Each Store should have their own Query, so if it's trying to query a (sub)store, then we should first call the store's Query and then modify the result, such as appending the StoreKey.Name().

BTW, just stating the obvious, please use types/errors.go as much as possible, and use sdk.Error.ABCICode() for the final implementation.

Reply:
So, each store must implement the abci query interface as well?
`Query(req abci.RequestQuery) (res abci.ResponseQuery)`

I was very hesistant about adding this to the interface, as it seemed a very complex piece to put into a very simple KVStore (complex meaning requires lots of semantics, many args in protobuf obj, etc.). The current Get/Set/Iterate is very clean and nice.

This was a work-around until I figured out how to get access to the stores to do proper queries, but if I should drag the abci Query into all the CommitStores, I can.